File: AppCpp/carpark_queueing_web.py

#Import Functions Required
import random
import time

arrivalrate = 1000
servicetime = 1000
spaces = 300
precision = 1

#Define Functions
def percentageoftime(percent,list):
    cumulativesum = 0
    totaltime = sum(list)
    targetsum = totaltime * percent / 100
    for index, value in enumerate(list):
        cumulativesum += value
        if cumulativesum >= targetsum:
            return index

def modelrun(arrivalrate,servicetime,spaces):

    #Intialise All Variables
    start_time = time.time()
    count_arrivals = 0
    count_carsparked = [0] * int(spaces+1)
    count_carsqueued = [0] * int(arrivalrate * servicetime)
    carsparked = []
    cyclecount = int(10000)
    arrival = 0
    carsqueued = 0
    queue = 0
    percentiles = [10,20,30,40,50,60,70,80,90,95,98,99]
    queuetime = 0
    queuetest = 0
    hours = 0
    minimumqueue = 0

    #Generate Arrivals
    
    for i in range (1,cyclecount * 3600 ):
        if i  > 3600*1000 and i%36000 ==0:
            if queuetest == round(carsqueued/count_arrivals,5) or i == 3000*3600:
                hours = i / 3600
                break
            else:
                queuetest = round(carsqueued/count_arrivals,5)
                
        # Check if new car arrived and at to carpark
        arrival = random.randint(1,3600 * precision)
        if arrival <= arrivalrate * precision:
            count_arrivals += 1
            if len(carsparked) < spaces:
                carsparked.append(servicetime)
            else: 
                queue +=1
                carsqueued +=1
                minimumqueue += carsparked[0]


        # Count current carpark utilisation
        count_carsparked[max(len(carsparked),0)] += 1
        count_carsqueued[queue] += 1
        queuetime += queue

        # Reduce parked cars time remaining by passing time
        if len(carsparked) > 0:
            carsparked = [item - 1 for item in carsparked]

            # move finished cars out and queued cars in    
            if carsparked[0] == 0:

                del carsparked[0]
                if queue > 0:
                    carsparked.append(servicetime)
                    queue -= 1

    cyclecount = hours

    for index, value in enumerate(count_carsparked):
        count_carsparked[index] = count_carsparked[index] / (hours * 3600)

    end_time = time.time()
    elapsed_time = end_time - start_time
    print("Model completed in ", int(round(elapsed_time,0)), " seconds", sep='')
    if hours == 3000:
        print("0.00001 stability not found. ",hours,"hours of survey data generated.")
    else:
        print("Stable solution found after",hours,"hours of survey data")
    print("Cars Queued = ", round(carsqueued * 100 / count_arrivals,2),"%", sep='')
    if carsqueued > 0:
        print("Average Queue time per Arrival/Queued Vehicle = ", round(queuetime / count_arrivals),"/",round(queuetime / carsqueued),"seconds")
    # print("Random Arrivals = ", round(count_arrivals / cyclecount / precision,1))
    # print("Service Time = ", servicetime)
    print("Perfect Arrivals  Demand = ", round(count_arrivals / cyclecount / precision * servicetime / 3600,2),"spaces")
    print("Random Arrivals Demand percentiles")
    for value in percentiles:
        print(value,"th - ",percentageoftime(value,count_carsparked)," parked and ", percentageoftime(value,count_carsqueued)," Queued", sep='')

modelrun(arrivalrate,servicetime,spaces)

File: static/carpark_queueing_web.js
var headroom = 0;
function percentageOfTime(percent, list) {
    let cumulativeSum = 0;
    const totalTime = list.reduce((sum, value) => sum + value, 0);
    const targetSum = (totalTime * percent) / 100;

    for (let index = 0; index < list.length; index++) {
        cumulativeSum += list[index];
        if (cumulativeSum >= targetSum) {
            return index;
        }
    }
    return 0;
}

function calculatePercentiles(countCarsParked, countCarsQueued) {
    const percentiles = [10, 20, 30, 40, 50, 60, 70, 80, 90, 95, 98, 99];
    const results = new Map();

    for (const p of percentiles) {
        results.set(p, {
            parked: percentageOfTime(p, countCarsParked),
            queued: percentageOfTime(p, countCarsQueued)
        });
    }

    return Object.fromEntries(results);
}

function runSimulation(inp) {
    const startTime = performance.now();
    var lastarrival = 0;
    const arrivalThreshold = inp.arrivalRate / (3600 * inp.precision - inp.arrivalRate * headroom * inp.precision);
    const state = {
        carsParked: [],
        carsQueued: 0,
        countArrivals: 0,
        countCarsParked: new Array(inp.spaces + 1).fill(0),
        countCarsQueued: new Array(Math.ceil(inp.arrivalRate * inp.serviceTime * 100)).fill(0),
        queue: 0,
        queueTest: 0,
        queueTime: 0
    };

    let hours = 0;
    const serviceTimeSteps = Math.ceil(inp.serviceTime * inp.precision);  // Adjust for decimal serviceTime

    for (let i = 1; i <= inp.cycles * inp.hours_in_steps; i++) {
        if (i > 3600 * 1000 * inp.precision && i % 36000 * inp.precision === 0) {
            hours = i / inp.hours_in_steps;
            const currentQueueRatio = state.carsQueued / state.countArrivals;
            if (Math.abs(state.queueTest - currentQueueRatio) <= 1e-6) {
                break;
            } else {
                state.queueTest = currentQueueRatio;
            }
        }

        // Check if a new car arrived and add to the car park
        if (lastarrival + headroom * inp.precision < i){
            if (Math.random() <= arrivalThreshold) {
                state.countArrivals++;
                lastarrival = i
                if (state.carsParked.length < inp.spaces) {
                    state.carsParked.push(serviceTimeSteps);
                } else {
                    state.queue++;
                    state.carsQueued++;

                }
            }
        }    

        // Count current car park utilization
        const index = Math.max(state.carsParked.length, 0);
        state.countCarsParked[index]++;
        state.countCarsQueued[state.queue]++;
        state.queueTime += state.queue;

        // Reduce parked cars' time remaining
        if (state.carsParked.length >= 1) {
            for (let j = 0; j < state.carsParked.length; j++) {
                state.carsParked[j]--;
            }

            // Move finished cars out and queued cars in
            if (state.carsParked[0] === 0) {

                state.carsParked.shift();
                if (state.queue > 0) {
                    state.carsParked.push(serviceTimeSteps);  // Adjust for decimal serviceTime

                    state.queue--;
                }
            }
        }
    }

    const elapsedMs = performance.now() - startTime;
    return {
        elapsedTime: `${Math.round(elapsedMs)}ms`,
        hours,
        PercentCarsQueued: ((state.carsQueued * 100) / state.countArrivals).toFixed(1),
        QueueTimePerQueuedVehicle: (state.queueTime / state.carsQueued / inp.precision).toFixed(1),
        QueueTimePerArrival: (state.queueTime / state.countArrivals / inp.precision).toFixed(1),
        theoreticalMinSpace: ((inp.arrivalRate * (inp.serviceTime)) / inp.hours_in_steps * inp.precision).toFixed(2),
        modelDemandSpace: (state.countArrivals * (inp.serviceTime) / hours / 3600).toFixed(2),
        percentiles: calculatePercentiles(state.countCarsParked, state.countCarsQueued)
    };
}

// Format results as a readable string
export function formatResults(results) {
    if (!results || typeof results !== 'object') {
        return 'Error: Invalid simulation results.';
    }

    let output = `Queueing Results:\n`;
    output += `-----------------------------\n`;
    output += `Elapsed Time: ${results.elapsedTime || 'N/A'}\n`;
    output += `Simulated Hours: ${results.hours || 'N/A'}\n`;
    output += `Theoretical Min Spaces: ${results.theoreticalMinSpace || 'N/A'}\n`;
    output += `Model Demand Spaces: ${results.modelDemandSpace || 'N/A'}\n`;
    output += `Percentage of Cars Queued: ${results.PercentCarsQueued || 'N/A'}%\n`;
    output += `Avg. Queue Time per Queued Vehicle: ${results.QueueTimePerQueuedVehicle || 'N/A'} sec\n`;
    output += `Avg. Queue Time per Arrival: ${results.QueueTimePerArrival || 'N/A'} sec\n`;
    output += `-----------------------------\n`;
    output += `Percentiles (Parked | Queued):\n`;

    // ✅ Ensure percentiles are correctly formatted
    if (results.percentiles && typeof results.percentiles === 'object') {
        for (const [percentile, values] of Object.entries(results.percentiles)) {
            output += `${percentile}%: ${values.parked ?? 'N/A'} | ${values.queued ?? 'N/A'}\n`;
        }
    } else {
        output += `Percentiles data not available.\n`;
    }
    return output;
}

// Main function to simulate the process
export function test(a, b, c, h) {
    let d =1;
    if ((a % 1 !== 0) || (b % 1 !== 0) || (h % 1 !== 0)  ) {
        d = 10;
    }
    headroom = h
    const inputs = {
        arrivalRate: Number(a),
        cycles: 3000,
        precision: d,
        hours_in_steps: 3600 * d,
        serviceTime: Number(b),
        spaces: Number(c)
        
    };

    if (inputs.arrivalRate * inputs.serviceTime / 3600 > inputs.spaces) {
        return 'Error: Arrival rate too high for available spaces.';
    } else {
        try {
            const results = formatResults(runSimulation(inputs));
            return results;
        } catch (error) {
            console.error('Error running simulation:', error);
        }
    }
}
